Add tests(minitest), clean-up rubocop offences

misdoro opened the pull request:
- Add Gemfile to the project
- Add rubocop with basic configuration
- Add tests that cover SdNotify.ready (and `.notify` as a consequence)
- Refactor/simplify `.notify` code a bit.

Fixes #2 

@agis What do you think of it? I tried to keep the things fairly simple.

Commit:
Add Gemfile

Commit:
Add rubocop as dependency

Commit:
Tests+Refactor SdNotify.ready

- simplify the method
- use block syntax with Addrinfo.unix(sock, :DGRAM).connect

misdoro commented:
Also @agis do you have any preference between circleci, travis or github actions for the pipeline?

Commit:
Test that `notify` returns the number of bytes sent

agis commented on sd_notify.gemspec:
```diff
@@ -9,4 +11,10 @@ Gem::Specification.new do |s|
   s.homepage    = "https://github.com/agis/ruby-sdnotify"
   s.license     = "MIT"
+
+  s.required_ruby_version = ">= 2.3.0"
```
Is there a reason we're restricted to 2.3.0 or later?

misdoro replied:
No real reason. In `teardown` in tests I used `&.` operator which is available only since 2.3, but I can definitely revert back to `@socket.close if @socket` syntax.

agis replied:
Yeah I don't think it's worth it. Let's support 2.3 and earlier versions.

agis commented on lib/sd_notify.rb:
```diff
@@ -104,21 +106,16 @@ def self.watchdog?
     sock = ENV["NOTIFY_SOCKET"]
 
-    return nil if !sock
+    return nil unless sock
```
:fire: I belong in the "no unless" camp :grimacing: :fire:

misdoro replied:
Removed

agis commented on lib/sd_notify.rb:
```diff
@@ -104,21 +106,16 @@ def self.watchdog?
     begin
-      sock = Addrinfo.unix(sock, :DGRAM).connect
-      connected = true
-      sock.close_on_exec = true
```
TIL that Ruby 2.0.0+ automatically sets this flag. Perhaps this should be stated in the relevant commit description, for future reference.

misdoro replied:
:thinking: I'm not actually sure I get in which context this flag was initially useful. Was it to close the socket if the error handler is forking a new process and does `exec` behind ?

Do you have a specific test case in mind?

agis replied:
Not sure either, but I guess it's the sane thing to do (not sure why CLOEXEC isn't the default when opening new sockets). The original implementation also [sets it](https://github.com/systemd/systemd/blob/v244/src/libsystemd/sd-daemon/sd-daemon.c#L479).

Also, from man pages of `sd_listen_fds(3)`:

> This function call will set the FD_CLOEXEC flag for all passed file descriptors to avoid further inheritance to children of the calling process

I guess you just don't want other threads spawned from the main process to mess with this fd.

agis replied:
On a second thought, maybe we could leave the explicit set of `close_on_exec = true` so that we don't depend on Ruby's defaults and that the code is more clear?

As for the test case, if we make it explicit I don't think it's worth it.

misdoro replied:
Yeah, why not.
At least it will educate people :)

misdoro replied:
Updated, with "kind of" meaningful commit message.

agis reviewed (commented):
That's great! Thanks! I left some minor comments.

Regarding the CI, we could try Github Actions. I only used Travis until now but I guess Github Actions might be simpler.

:clap: :100: :tada:

misdoro committed:
Do not require ruby 2.3 minimum

misdoro committed:
Do not use unless

misdoro committed:
Add basic travis.yml

Test against rubies 2.2 to 2.7 and ruby-head

misdoro commented:
Stick to Travis for now: github actions do not support ruby versions before 2.4, and I would prefer to be testing against old rubies too. All builds are green: https://travis-ci.org/misdoro/ruby-sdnotify

Commit:
Force close_on_exec = true on socket

Explicitly setting it to be absolutely sure no forks of the process will ever keep the file descriptor open.
See e.g. https://www.gnu.org/software/libc/manual/html_node/Descriptor-Flags.html for more info.

agis commented:
That's great! Thanks @misdoro! :clap: :100:
